Source/JNA/waffle-jna/src/main/java/waffle/jaas/WindowsLoginModule.java

```diff
@@ -160,12 +168,18 @@ public boolean login() throws LoginException {
             }
 
             this.principals = new LinkedHashSet<>();
+            Set<String> domainlist = new LinkedHashSet<>();
+            if (this.domains != null)
+                domainlist.addAll(Arrays.asList(this.domains.split(",")));
             // add the main user principal to the subject principals
             this.principals.addAll(WindowsLoginModule.getUserPrincipals(windowsIdentity, this.principalFormat));
-            if (this.roleFormat != PrincipalFormat.NONE) {
+            if ((this.roleFormat != PrincipalFormat.NONE) || (!domainlist.isEmpty())) {
                 // add Windows Groups as role principles
                 for (final IWindowsAccount group : windowsIdentity.getGroups()) {
                     this.principals.addAll(getRolePrincipals(group, this.roleFormat));
+                    if (domainlist.contains(group.getDomain())) {
+                        this.principals.add(new RolePrincipal(group.getName()));
+                    }
```

Comment:
Shouldn't `getRolePrincipals` be responsible for adding those instead of doing it here?

Reply:
It would make sense to move it GetRolePrincipals.  This method would then either not be static anymore, or require the domains to be passed as parameter (which kind of defeats the object of making it static).  The better implementation will move lines 170 to 184 into a seperate method.  I left it here as this is where the previous GroupPrincipal was also added.  

In essence we have an WindowsIdentity which has Groups which is a set of Account.  This needs to be translated into Java Principles.  In essence we have UserPrincipals, GoupPrincipals (not the previous waffle GroupPrincipal, but a Principal representing the Windows Group in a domain) and RolePrincipal, which would be the security roles required by the application.  Somewhere Windows groups have to be mapped to application roles.

A transformation method takes a WindowsIdentity and generates the Principals for a Subject from this identity.  This includes generating the UserPrincipal and the one or more translations/mapings from the Groups to Roles.  The Principles describe the "identities" of the WindowsIdentity which includes role and/or group memberships. 

Doing it correctly would then actually move the creation of all Principals (or for the jna-tomcat Windowsrealm the Strings (Names/Identities) from a single WindowsIdentity into a seperate method.

A transformation method could in theory be re-used in JAAS, and the various jna-tomcat* implementations.

We could implement different LoginModules for different types of mapping or make them options on one LoginModule.

What I have implemented now, makes it possible to interchange the Tomcat UserDatabaseRealm, DataSourceRealm and JAASRealm as realms for the same application which uses Tomcat for authentication and passes the roles to a web-app, or for Tomcat itself.  Without the modification, this is no possible,